### PR TITLE
Support DataFrame `read_csv`

### LINKED ISSUE #367: `read_csv` support for DataFrame
**Is your feature request related to a problem? Please describe.**

Add `mars.dataframe.read_csv` method to support read csv files.

### LINKED ISSUE #417: FileSystem interface
**Is your feature request related to a problem? Please describe.**

We need a `FileSystem` interface which provides the basic capabilities to operate on external storage, such as HDFS, S3 or oss etc. It could be an interface, and delegate the implementations to different libraries like [pyarrow.HadoopFileSystem](https://arrow.apache.org/docs/python/filesystems.html#hadoop-file-system-hdfs). 

Also, we should support parsing the uri which contains protocol such as `hdfs://` and other things correctly.

This would be useful if we are about to save our data to external storage with a unified way.

**Additional context**

I am searching out solution for `zarr` support of tensor and `read_csv` support of DataFrame, and I realized that they all require the `FileSystem` if we want to read or write data to external storage.

### PR DESCRIPTION
**Pull request by hekaisheng**
## What do these changes do?

Support create DataFrame from a single csv files.

## Related issue number

resolve #367.
resolve #417

### CODE REVIEW

File: mars/dataframe/datasource/read_csv.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+                                     index_value=df.index_value,
+                                     columns_value=df.columns,
+                                     chunks=out_chunks, nsplits=nsplits)
+
+    @classmethod
+    def execute(cls, ctx, op):
+        out_df = op.outputs[0]
+        with open_file(op.path, compression=op.compression, storage_options=op.storage_options) as f:
+            if op.compression is not None:
+                df = pd.read_csv(BytesIO(f.read()), header=op.header, names=op.names, index_col=op.index_col,
```
Comment (qinxuye) [inline mars/dataframe/datasource/read_csv.py]: Use `cudf` or `pd` according to `gpu` property.

File: mars/dataframe/datasource/read_csv.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    if isinstance(mini_df.index, pd.RangeIndex):
+        index_value = parse_index(pd.RangeIndex(0))
+    else:
+        index_value = parse_index(mini_df.index)
+    columns_value = parse_index(mini_df.columns, store_data=True)
+    if index_col and not isinstance(index_col, int):
+        index_col = list(mini_df.columns).index(index_col)
+    names = list(mini_df.columns)
+    op = DataFrameReadCSV(path=path, names=names, sep=sep, header=header or 0, index_col=index_col,
+                          compression=compression, storage_options=storage_options, **kwargs)
```
Comment (qinxuye) [inline mars/dataframe/datasource/read_csv.py]: `kwargs` is not used in execution.

File: mars/dataframe/datasource/read_csv.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+                if end == start:
+                    # the last chunk may be empty
+                    df = build_empty_df(out_df.dtypes)
+                else:
+                    if out_df.index == (0, 0):
+                        # The first chunk contains header
+                        df = pd.read_csv(b, header=op.header, names=op.names, index_col=op.index_col,
+                                         dtype=out_df.dtypes.to_dict())
+                    else:
+                        df = pd.read_csv(b, names=op.names, index_col=op.index_col,
```
Comment (qinxuye) [inline mars/dataframe/datasource/read_csv.py]: Could be refined  a bit to reduce duplication.

File: mars/dataframe/datasource/read_csv.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+
+def read_csv(path, names=None, sep=None, index_col=None, compression=None, header=None, dtype=None,
+             chunk_bytes=None, storage_options=None, **kwargs):
+    # infer dtypes and columns
+    with open_file(path, compression=compression, storage_options=storage_options) as f:
+        if header:
+            _ = [f.readline() for _ in range(header)]  # noqa: F841
+        head = f.readline()
+        first_row = f.readline()
```
Comment (qinxuye) [inline mars/dataframe/datasource/read_csv.py]: If the first row contains some null data, data types may not be inferred.

File: mars/dataframe/datasource/read_csv.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+                                         dtype=out_df.dtypes.to_dict())
+        ctx[out_df.key] = df
+
+    def __call__(self, index_value=None, columns_value=None, dtypes=None, chunk_bytes=None):
+        shape = (np.nan, len(dtypes))
+        return self.new_dataframe(None, shape, dtypes=dtypes, index_value=index_value,
+                                  columns_value=columns_value, chunk_bytes=chunk_bytes)
+
+
+def read_csv(path, names=None, sep=None, index_col=None, compression=None, header=None, dtype=None,
```
Comment (qinxuye) [inline mars/dataframe/datasource/read_csv.py]: I think multi files could be supported as well.

File: mars/dataframe/datasource/tests/test_datasource_execution.py [hunk 6]
```diff
@@ -130,3 +133,66 @@ def testFromRecordsExecution(self):
         df2 = from_records(ndarr)
         df2_result = self.executor.execute_dataframe(df2, concat=True)[0]
         pd.testing.assert_frame_equal(df2_result, pdf_expected)
+
+    def testReadCSVExecution(self):
```
Comment (qinxuye) [inline mars/dataframe/datasource/tests/test_datasource_execution.py]: Null data could be tested.

File: mars/filesystem.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            stat['type'] = 'other'
+        return stat
+
+
+file_systems = {
+    'file': LocalFileSystem
+}
+
+
+def get_fs(path, storage_options):
```
Comment (qinxuye) [inline mars/filesystem.py]: HDFS filesystem should be added, test required as well.

File: mars/filesystem.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+from .compat import urlparse
+
+
+compressions = {
+    'gzip': lambda f: GzipFile(fileobj=f)
+}
+
+if lz4:
+    compressions['lz4'] = lz4.frame.open
```
Comment (qinxuye) [inline mars/filesystem.py]: How about other compress types, e.g. snappy, zstd

File: mars/filesystem.py (hunk 7 above)
Comment (hekaisheng) [reply mars/filesystem.py]: As HDFS tests need extra CI settings, I will add tests in next PR.

File: mars/filesystem.py (hunk 8 above)
Comment (hekaisheng) [reply mars/filesystem.py]: I will add more compress types and add tests for file system next time.

File: mars/dataframe/datasource/read_csv.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from io import BytesIO
+
+import pandas as pd
+import numpy as np
+try:
+    import cudf
```
Comment (qinxuye) [inline mars/dataframe/datasource/read_csv.py]: Use `lazy_import` for `cupy`.

**Pull request merged**